### PR TITLE
SQLA2: avoid unnecessary locking when working with TIs

### PR DESCRIPTION
**Pull request by Dev-iL**
Related: #59218 

In SQLAlchemy 2.0, we need to explicitly tell `get()` not to issue a new `SELECT` for an already-loaded object. Using `with_for_update=True` tells it to use the locked version.

All changes are backward-compatible with SQLAlchemy 1.4:
- The with_for_update=True parameter exists in 1.4 and works correctly
- Passing a string to TaskReschedule constructor works in both versions
- Skipping fail_fast in edge cases doesn't affect normal operation

These changes fix the `test_ti_update_state_reschedule_mysql_limit` test which would otherwise time out:
image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

### CODE REVIEW

**Comment by Dev-iL** [comment]
Does anyone think this PR requires additional tags so more tests run?
cc: @potiuk @vincbeck

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Nice

File: airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py [hunk 1]
```diff
@@ -548,22 +548,22 @@ def _create_ti_state_update_query_and_update_state(
                 if session.bind is not None:
                     query = TI.duration_expression_update(timezone.utcnow(), query, session.bind)
                 query = query.values(state=TaskInstanceState.FAILED)
-                ti = session.get(TI, ti_id_str)
-                if ti is not None:
-                    _handle_fail_fast_for_dag(ti=ti, dag_id=dag_id, session=session, dag_bag=dag_bag)
```
Comment (kaxil) [inline airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: @Dev-iL Why was this silently dropped?

Comment (Dev-iL) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: Possibly because it was not covered by tests and therefore unneeded

Comment (kaxil) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: Not covered by test isn't a good reason to drop things

Comment (kaxil) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: So was this an intentional change or unintentional?

Comment (Dev-iL) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: It looks unintentional 6mo later, but perhaps it was resolving some flakiness or inconsistency between sqla versions, or deemed unnecessary in light of the other changes. 

The lack of coverage is not the motivation for the removal, but rather the reason it "succeeded".

Comment (Dev-iL) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: Re the "silently" part - the PR description does say that

> Skipping fail_fast in edge cases doesn't affect normal operation

Comment (kaxil) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: fail_fast is a feature that stopped working for reschedule cases.

Comment (kaxil) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: > Skipping fail_fast in edge cases doesn't affect normal operation

That's completely wrong. Normal operation isn't things just working. fail_fast is a feature not an edge case!

Comment (kaxil) [reply airflow-core/src/airflow/api_fastapi/execution_api/routes/task_instances.py:553]: Will create a PR to restore that.
